File: fs/Practice2/DynamicProgramming/Other/LongestIncreasingSubsequence.fs
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=DPL_1_D

stdin.ReadLine ()
|> int
|> fun n ->
  (
    List.init n (fun _ -> stdin.ReadLine () |> int),
    Array.create (n + 1) -1
  )
  |> fun (ns, memo) ->
    ns
    |> List.fold (
      fun len x ->
        if memo.[len] < x
        then
          (memo.[len + 1] <- x)
          |> fun _ -> len + 1
        else
          memo
          |> Array.findIndex ((<=) x)
          |> fun i -> memo.[i] <- x
          |> fun _ -> len
    ) 0
|> printfn "%d"

// input:
// 5
// 5
// 1
// 3
// 2
// 4

// memo:
// [|-; 1; 2; 4; -; -|]

// output:
// 3

File: scala/Practice2/DynamicProgramming/Other/LongestIncreasingSubsequence.scala
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=DPL_1_D

object Main {

  implicit class ScopeFunction[T](x: T) {
    def let[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    io.StdIn.readLine
      .toInt
      .let(n =>
        (
          List.range(0, n).map(_x => io.StdIn.readLine.toInt),
          Array.fill(n + 1)(-1)
        )
        .let { case (ns, memo) =>
          ns.foldLeft(0) { case (len, x) =>
            if (memo(len) < x)
              (memo(len + 1) = x)
                .let(_x => len + 1)
            else
              memo.indexWhere(_ >= x)
                .let(i => memo(i) = x)
                .let(_x => len)
          }
        }
      )
      .let(println)
  )
}

// https://onlinejudge.u-aizu.ac.jp/recent_judges/DPL_2_A/judge/5320964/dsudo/Scala
